datafusion/src/optimizer/single_distinct_to_groupby.rs

```diff
@@ -0,0 +1,278 @@
+///   - Aggregation
+///          GROUP BY (k)
+///          F1(DISTINCT s)
+///       - X
+///
+///   into
+///
+///   - Aggregation
+///            GROUP BY (k)
+///            F1(s)
+///       - Aggregation
+///               GROUP BY (k, s)
+///            - X
+///   </pre>
+///   <p>
+pub struct SingleDistinctToGroupBy {}
+
+impl SingleDistinctToGroupBy {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
+    match plan {
+        LogicalPlan::Aggregate {
+            input,
+            aggr_expr,
+            schema,
+            group_expr,
+        } => {
+            if is_single_distinct_agg(plan) {
+                let mut all_group_args = Vec::with_capacity(group_expr.len());
+                all_group_args.append(&mut group_expr.clone());
+                // remove distinct and collection args
+                let new_aggr_expr = aggr_expr
+                    .iter()
+                    .map(|agg_expr| match agg_expr {
+                        Expr::AggregateFunction { fun, args, .. } => {
+                            all_group_args.append(&mut args.clone());
+                            Expr::AggregateFunction {
+                                fun: fun.clone(),
+                                args: args.clone(),
+                                distinct: false,
+                            }
+                        }
+                        _ => agg_expr.clone(),
+                    })
+                    .collect::<Vec<_>>();
+
+                let all_field = all_group_args
+                    .iter()
+                    .map(|expr| expr.to_field(input.schema()).unwrap())
+                    .collect::<Vec<_>>();
+
+                let grouped_schema = DFSchema::new(all_field).unwrap();
+                let grouped_agg = LogicalPlan::Aggregate {
+                    input: input.clone(),
+                    group_expr: all_group_args,
+                    aggr_expr: Vec::new(),
+                    schema: Arc::new(grouped_schema.clone()),
+                };
+                let grouped_agg = optimize_children(&grouped_agg);
+                let final_agg_schema = Arc::new(
+                    DFSchema::new(
+                        group_expr
+                            .iter()
+                            .chain(new_aggr_expr.iter())
+                            .map(|expr| expr.to_field(&grouped_schema).unwrap())
+                            .collect::<Vec<_>>(),
+                    )
+                    .unwrap(),
+                );
+
+                let final_agg = LogicalPlan::Aggregate {
+                    input: Arc::new(grouped_agg.unwrap()),
+                    aggr_expr: new_aggr_expr,
+                    schema: final_agg_schema.clone(),
+                    group_expr: group_expr.clone(),
+                };
+
+                let mut alias_expr: Vec<Expr> = Vec::new();
```

Comment:
It might help here to explain the rationale for adding this alias (so the aggregates are displayed in the same way even after the rewrite). It is important but may not be obvious to other readers